### PR TITLE
Implement support for reactive generator functions

### PR DESCRIPTION
**Pull request by philippjfr**
In looking at https://github.com/holoviz/panel/issues/5016 I realized that the only way to write reactive expressions with multiple steps in a natural way would be using generators. So I had a go at implementing support for them and now I'm in love, here's a simplified version of the app in #5016:

### Asynchronous Generator

```python
import asyncio
import panel as pn

pn.extension(sizing_mode="stretch_width", template='bootstrap')

run = pn.widgets.Button(name="Press to run calculation")

async def runner(running):
    if not running:
        yield "Calculation did not run yet"
        return
    yield pn.Column(
        pn.indicators.LoadingSpinner(value=True, width=25, height=25),
        "Running...Please wait!",
    )
    time_start = time.perf_counter()
    await asyncio.sleep(1.5)
    time_end = time.perf_counter()
    yield f"""Done!
    The function took {time_end - time_start:1.1f} seconds to complete"""

pn.Column(
    "# Calculation Runner", run, pn.bind(runner, run)
).servable()
```

### Classic Generator

```python
import time
import panel as pn

pn.extension(sizing_mode="stretch_width", template='bootstrap')

run = pn.widgets.Button(name="Press to run calculation")

def runner(running):
    if not running:
        yield "Calculation did not run yet"
        return
    yield pn.Column(
        pn.indicators.LoadingSpinner(value=True, width=25, height=25),
        "Running...Please wait!",
    )
    time_start = time.perf_counter()
    time.sleep(1.5)
    time_end = time.perf_counter()
    yield f"""Done!
    The function took {time_end - time_start:1.1f} seconds to complete"""

pn.Column(
    "# Calculation Runner", run, pn.bind(runner, run)
).servable()
```

generator

## Streaming example

```python
import asyncio
import panel as pn
import random

pn.extension()

@pn.depends()
async def stream():
    while True:
        await asyncio.sleep(0.1)
        yield random.randint(0, 100)
        
pn.indicators.Number(value=stream, name='Random')
```

random_generator

### CODE REVIEW

**Comment by MarcSkovMadsen** [comment]
The streaming example is extremely powerful. The Async and Classic generators too, but here we have to teach our users `yield`.

**Comment by MarcSkovMadsen** [comment]
Does `pn.bind(runner, run)` run twice when I click the `run` button? Has it always done that?

File: panel/depends.py [hunk 1]
```diff
@@ -216,12 +223,24 @@ def eval_fn():
                 fn = eval_function(p)
         return fn
 
-    if iscoroutinefunction(function):
+    if inspect.isasyncgenfunction(function):
```
Comment (MarcSkovMadsen) [inline panel/depends.py:226]: use `from inspect import isasyncgenfunction` instead. You will save an attribute lookup each time this line is run which could be thousands or millions of times.

**Comment by MarcSkovMadsen** [comment]
I added another use case for generator functions in Panel that I believe should be supported as well. Please consider this in the context of this PR. Thanks. https://github.com/holoviz/panel/issues/5020

**Comment by philippjfr** [comment]
>Does pn.bind(runner, run) run twice when I click the run button? Has it always done that?

It runs once but if you `bind` a Button to a function it'll run the first time with a value of False and on every click it'll run with a value of `True`.

>Does this work with other widgets like an IntSlider?

Absolutely, it works just like a regular function with `bind`.

>Does the streaming example perform nicely if you have 100 streaming generators? I'm asking because I've found that its not performant to setup 100 period callbacks - its better to update 100 times in one. It would be so great not to have to couple the streaming components.

Absolutely, this should be significantly more lightweight than periodic callbacks, here's 100 instances of the streaming generator

many_generators

File: panel/depends.py (hunk 1 above)
Comment (philippjfr) [reply panel/depends.py:226]: I'll do that but note that this is not inside the core loop and won't be called on each update.

**Comment by philippjfr** [comment]
Okay, feature-wise this is now done. Last item was implementing synchronization primitives to ensure the async generators were updated when a dependency changes, e.g. we make the timeout dynamic:

```python
import asyncio
import panel as pn
import random

pn.extension()

timeout = pn.widgets.FloatSlider(start=0.1, end=1.0, value=1, name='Timeout')

@pn.depends(timeout)
async def stream(timeout):
    while True:
        await asyncio.sleep(timeout)
        yield random.randint(0, 100)

pn.Column(
    timeout,
    pn.FlexBox(*(pn.indicators.Number(value=stream, name='Random', font_size='12pt') for _ in range(200)))
).servable()
```

async_gen_deps

**Pull request merged**